File: docs/eos/examples/compliance-starter/dsar.go
// dsar.go — Data Subject Access Requests: export (access/portability) + erase (right to erasure).
// Go parallel of dsar.mjs / dsar.py. GDPR Art.15/17/20 · PIPL 45/47. Landing point: data-api
// "Support subject deletion / export where PII is stored" + "Every deletion … is audited".
//
// A Source is any store that can export/erase ONE subject's rows — register one adapter per
// table/service. This is the pluggable "your data" seam. Handlers are synchronous here for a
// zero-dep skeleton; wrap them in goroutines / a task queue if exports are large.
package compliance

// Source exports/erases one subject's rows for a single table or service.
type Source interface {
	Name() string
	Export(subjectID string) any // structured rows for this subject (access/portability)
	Erase(subjectID string) int  // returns how many rows were erased
}

// ExportBundle is the structured, machine-readable access/portability payload (GDPR Art.20).
type ExportBundle struct {
	SubjectID   string
	GeneratedAt string
	Data        map[string]any
}

// ExportSubject gathers every source's rows for the subject and audits the access.
// Pass nil for log to use a fresh trail.
func ExportSubject(subjectID string, sources []Source, log *AuditLog) ExportBundle {
	if log == nil {
		log = NewAuditLog()
	}
	data := map[string]any{}
	names := make([]any, 0, len(sources))
	for _, s := range sources {
		data[s.Name()] = s.Export(subjectID)
		names = append(names, s.Name())
	}
	log.Record(Entry{"type": "dsar.export", "subjectId": subjectID, "sources": names})
	return ExportBundle{SubjectID: subjectID, GeneratedAt: now(), Data: data}
}

// EraseReceipt records WHAT was erased WHERE (not the data itself) — safe to keep/audit.
type EraseReceipt struct {
	SubjectID string
	ErasedAt  string
	Reason    string
	Erased    map[string]int
}

// EraseSubject runs erasure on every source and audits it. An empty reason defaults to
// "dsar.erasure". Backups inherit the same rule: re-run erasure on restore, or document the
// retention window + encryption so restored data self-expires. (.github/instructions/data-api)
func EraseSubject(subjectID string, sources []Source, reason string, log *AuditLog) EraseReceipt {
	if log == nil {
		log = NewAuditLog()
	}
	if reason == "" {
		reason = "dsar.erasure"
	}
	erased := map[string]int{}
	for _, s := range sources {
		erased[s.Name()] = s.Erase(subjectID)
	}
	log.Record(Entry{"type": "dsar.erase", "subjectId": subjectID, "reason": reason, "erased": erased})
	return EraseReceipt{SubjectID: subjectID, ErasedAt: now(), Reason: reason, Erased: erased}
}

File: docs/eos/examples/compliance-starter/dsar.py
"""dsar.py — Data Subject Access Requests: export (access/portability) + erase (right to erasure).

GDPR Art.15/17/20 · PIPL 45/47. Landing point: data-api "Support subject deletion / export
where PII is stored (GDPR/CCPA-style 'right to erasure/access')" + "Every deletion … is audited".

A ``source`` is any object exposing ``name`` + ``export(subject_id)`` + ``erase(subject_id)`` —
register one adapter per table/service. This is the pluggable "your data" seam.
Handlers are synchronous here for a zero-dep stdlib skeleton; wrap in your framework's
async/await or a task queue if exports are large.
"""
from __future__ import annotations

from datetime import datetime, timezone
from typing import Any, Protocol

from audit import AuditLog, audit


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


class Source(Protocol):
    name: str

    def export(self, subject_id: str) -> Any: ...

    def erase(self, subject_id: str) -> int: ...


def export_subject(subject_id: str, sources: list[Source], log: AuditLog | None = None) -> dict:
    """Access + Portability: structured, machine-readable bundle (GDPR Art.20)."""
    log = log if log is not None else audit()
    data = {s.name: s.export(subject_id) for s in sources}
    log.record(type="dsar.export", subjectId=subject_id, sources=[s.name for s in sources])
    return {"subjectId": subject_id, "generatedAt": _now(), "data": data}


def erase_subject(subject_id: str, sources: list[Source], *, reason: str = "dsar.erasure",
                  log: AuditLog | None = None) -> dict:
    """Right to erasure. Returns an audit-friendly record of WHAT was erased WHERE (not the data)."""
    log = log if log is not None else audit()
    erased = {s.name: s.erase(subject_id) for s in sources}
    # Backups inherit the same rule: re-run erasure on restore, or document the retention window
    # and encryption so restored data self-expires. (.github/instructions/data-api)
    log.record(type="dsar.erase", subjectId=subject_id, reason=reason, erased=erased)
    return {"subjectId": subject_id, "erasedAt": _now(), "reason": reason, "erased": erased}
